### PR TITLE
WT-2592 Fix joins for the non-recno, non-raw case

### PR DESCRIPTION
**Pull request by ddanderson**
This fixes a rather fundamental problem.  Any joins written in C that do not use recno were not seeing results.  Internally we are using raw keys, but our final 'main' cursor may not be in raw mode, so we may not be able to retrieve the values using the key we have.

Unfortunately, the bulk of the test cases for join are in Python, which use RAW mode, so unaffected.  Our C example uses recno, which took a different path.

### CODE REVIEW

**Comment by ddanderson** [comment]
@michaelcahill, can you review please?

File: src/cursor/cur_join.c [hunk 1]
```diff
@@ -1104,19 +1103,21 @@ __curjoin_next(WT_CURSOR *cursor)
...
-			WT_ERR(__wt_vunpack_uint(&p, iter->curkey->size,
-			    &cjoin->iface.recno));
-			c->set_key(c, cjoin->iface.recno);
-		} else
-			c->set_key(c, iter->curkey);
-		WT_ERR(c->search(c));
+		F_SET(c, WT_CURSTD_RAW);
+		c->set_key(c, iter->curkey);
+		if (!F_ISSET(cursor, WT_CURSTD_RAW))
+			F_CLR(c, WT_CURSTD_RAW);
```
Comment (michaelcahill) [inline src/cursor/cur_join.c]: @ddanderson, we have `__wt_cursor_set_raw_key` that should take care of these 4 lines for you:
```c
    __wt_cursor_set_raw_key(c, iter->curkey);
```

**Comment by michaelcahill** [comment]
@ddanderson, it sounds like we should we add a non-recno join test in C?

Otherwise, this looks fine to me with a minor simplification.

**Comment by ddanderson** [comment]
@michaelcahill, thanks for the simplification!  I added WT-2605 to track the need for better tests.  That should be easier once Alex has a C test framework (WT-2554).

Merging now.

**Pull request merged**